modules/configuration/src/main/java/org/apache/ignite/configuration/internal/NamedListConfiguration.java

```diff
@@ -38,89 +44,41 @@
+        BiFunction<List<String>, String, T> creator) {
+        super(prefix, key, rootKey, changer);
         this.creator = creator;
     }
 
-    /**
-     * Copy constructor.
-     * @param base Base to copy from.
-     * @param configurator Configurator to attach to.
-     * @param root Root of the configuration.
-     */
-    private NamedListConfiguration(
-        NamedListConfiguration<VIEW, T, INIT, CHANGE> base,
-        Configurator<? extends DynamicConfiguration<?, ?, ?>> configurator,
-        DynamicConfiguration<?, ?, ?> root
-    ) {
-        super(base.prefix, base.key, false, configurator, root);
-
-        this.creator = base.creator;
-
-        for (Map.Entry<String, T> entry : base.values.entrySet()) {
-            String k = entry.getKey();
-            T value = entry.getValue();
-
-            final T copy = (T) ((DynamicConfiguration<VIEW, INIT, CHANGE>) value).copy(root);
-            add(copy);
-
-            this.values.put(k, copy);
-        }
-    }
-
-    /** {@inheritDoc} */
-    @Override public void init(NamedList<INIT> list) {
-        list.getValues().forEach((key, init) -> {
-            if (!values.containsKey(key)) {
-                final T created = creator.apply(qualifiedName, key);
-                add(created);
-                values.put(key, created);
-            }
-
-            values.get(key).init(init);
-        });
-    }
-
     /**
      * Get named configuration by name.
      * @param name Name.
      * @return Configuration.
      */
     public T get(String name) {
-        return values.get(name);
-    }
+        refreshValue();
 
-    /** {@inheritDoc} */
-    @Override public NamedList<VIEW> value() {
-        return new NamedList<>(values.entrySet().stream().collect(Collectors.toMap(Map.Entry::getKey, it -> it.getValue().value())));
+        return values.get(name); //TODO Exceptions.
     }
 
     /** {@inheritDoc} */
-    @Override public void changeWithoutValidation(NamedList<CHANGE> list) {
-        list.getValues().forEach((key, change) -> {
-            if (!values.containsKey(key)) {
-                final T created = creator.apply(qualifiedName, key);
-                add(created);
-                values.put(key, created);
-            }
+    @Override protected synchronized void refreshValue0(NamedListView<VIEW> newValue) {
+        //TODO Just swap it, we don't need actual concurrent access.
+        Set<String> newKeys = newValue.namedListKeys();
 
-            values.get(key).changeWithoutValidation(change);
-        });
-    }
+        values.keySet().removeIf(key -> !newKeys.contains(key));
 
-    /** {@inheritDoc} */
-    @Override public NamedListConfiguration<VIEW, T, INIT, CHANGE> copy(DynamicConfiguration<?, ?, ?> root) {
-        return new NamedListConfiguration<>(this, configurator, root);
+        for (String newKey : newKeys) {
+            if (!values.containsKey(newKey))
+                values.put(newKey, creator.apply(keys, newKey));
```

Comment:
Doesn't this create default values?

Reply:
See no problems with that ;) These defaults couldn't be read because of "refresh" method guarantees.